### ISSUE TITLE
Python dependencies

### ISSUE THREAD

**Issue by fdila** [MISTLab/Swarm-SLAM#51]
Hi, 
I'm trying to run your code but I'm having some issues with the dependencies.
I'm using the provided Dockerfile in the cslam_experiments repo, however I'm still encountering issues with numpy.
This is the error I get when I try to launch the `cslam_lidar.launch.py
`
```
[INFO] [launch]: All log files can be found below /root/.ros/log/2024-11-26-11-36-37-224912-lmmartinez-INVETT-2984504
[INFO] [launch]: Default logging verbosity is set to INFO
[INFO] [loop_closure_detection_node.py-1]: process started with pid [2984505]
[INFO] [lidar_handler_node.py-2]: process started with pid [2984507]
[INFO] [pose_graph_manager-3]: process started with pid [2984509]
[pose_graph_manager-3] [INFO] [1732620997.263394780] [r0.cslam_pose_graph_manager]: Visualization enabled.
[pose_graph_manager-3] [INFO] [1732620997.266485814] [r0.cslam_pose_graph_manager]: Initialization done.
[lidar_handler_node.py-2] [INFO] [1732620998.245610024] [r0.cslam_map_manager]: Initialization done.
[loop_closure_detection_node.py-1] [INFO] [1732620998.356071917] [r0.cslam_loop_closure_detection]: Using ScanContext.
[loop_closure_detection_node.py-1] 
[loop_closure_detection_node.py-1] A module that was compiled using NumPy 1.x cannot be run in
[loop_closure_detection_node.py-1] NumPy 2.0.2 as it may crash. To support both 1.x and 2.x
[loop_closure_detection_node.py-1] versions of NumPy, modules must be compiled with NumPy 2.0.
[loop_closure_detection_node.py-1] Some module may need to rebuild instead e.g. with 'pybind11>=2.12'.
[loop_closure_detection_node.py-1] 
[loop_closure_detection_node.py-1] If you are a user of the module, the easiest solution will be to
[loop_closure_detection_node.py-1] downgrade to 'numpy<2' or try to upgrade the affected module.
[loop_closure_detection_node.py-1] We expect that some modules will need time to support NumPy 2.
[loop_closure_detection_node.py-1] 
[loop_closure_detection_node.py-1] Traceback (most recent call last):  File "/Swarm-SLAM/install/cslam/lib/cslam/loop_closure_detection_node.py", line 107, in <module>
[loop_closure_detection_node.py-1]     lcd = LoopClosureDetection()
[loop_closure_detection_node.py-1]   File "/Swarm-SLAM/install/cslam/lib/cslam/loop_closure_detection_node.py", line 97, in __init__
[loop_closure_detection_node.py-1]     self.glcd = GlobalDescriptorLoopClosureDetection(
[loop_closure_detection_node.py-1]   File "/Swarm-SLAM/install/cslam/local/lib/python3.10/dist-packages/cslam/global_descriptor_loop_closure_detection.py", line 142, in __init__
[loop_closure_detection_node.py-1]     import cv2
[loop_closure_detection_node.py-1] AttributeError: _ARRAY_API not found
[loop_closure_detection_node.py-1] Traceback (most recent call last):
[loop_closure_detection_node.py-1]   File "/Swarm-SLAM/install/cslam/lib/cslam/loop_closure_detection_node.py", line 107, in <module>
[loop_closure_detection_node.py-1]     lcd = LoopClosureDetection()
[loop_closure_detection_node.py-1]   File "/Swarm-SLAM/install/cslam/lib/cslam/loop_closure_detection_node.py", line 97, in __init__
[loop_closure_detection_node.py-1]     self.glcd = GlobalDescriptorLoopClosureDetection(
[loop_closure_detection_node.py-1]   File "/Swarm-SLAM/install/cslam/local/lib/python3.10/dist-packages/cslam/global_descriptor_loop_closure_detection.py", line 142, in __init__
[loop_closure_detection_node.py-1]     import cv2
[loop_closure_detection_node.py-1] ImportError: numpy.core.multiarray failed to import
[ERROR] [loop_closure_detection_node.py-1]: process has died [pid 2984505, exit code 1, cmd '/Swarm-SLAM/install/cslam/lib/cslam/loop_closure_detection_node.py --ros-args -r __node:=cslam_loop_closure_detection -r __ns:=/r0 --params-file /Swarm-SLAM/install/cslam_experiments/share/cslam_experiments/config/kitti_lidar.yaml --params-file /tmp/launch_params_wahswam5'].

```
Do you have any insight on what could be done to solve this?

**Comment by coisi-s** [comment]
1. The Problem
NumPy Version Mismatch: The error message clearly states that a module was compiled using NumPy 1.x and is now being run with NumPy 2.0.2, which may cause crashes. Some modules need to be recompiled with NumPy 2.0 to support both 1.x and 2.x versions properly.
cv2 Import Issue: There's also an issue when trying to import cv2. It seems that the _ARRAY_API is not found, and further down, it shows that numpy.core.multiarray failed to import. This could be related to the NumPy version problem or potentially an issue with the installation or configuration of cv2 itself.
2. Possible Solutions
Option 1: Downgrade NumPy
Explanation: As suggested in the error message, one straightforward solution could be to downgrade NumPy to a version less than 2.0. This would ensure that the modules compiled with NumPy 1.x can run without issues related to the version mismatch.
Steps:
First, identify the virtual environment (if any) where your project is running. If you're using Docker, make sure you're in the correct container environment.
Then, use the package manager (usually pip) to uninstall the current version of NumPy. For example, if you're in a virtual environment activated by source activate myenv (replace myenv with your actual environment name), you would run:
bash
复制
pip uninstall numpy
Next, install a version of NumPy less than 2.0. You can specify the version explicitly. For example, to install NumPy 1.24.3:
bash
复制
pip install numpy==1.24.3
Option 2: Recompile the Affected Modules
Explanation: The error mentions that some modules may need to be recompiled with pybind11>=2.12 to support both NumPy 1.x and 2.x versions. This option is a bit more involved but might be necessary if downgrading NumPy is not a viable solution for your specific use case.
Steps:
First, ensure that pybind11>=2.12 is installed in your environment. If not, install it using pip:
bash
复制
pip install pybind11>=2.12
Then, you'll need to identify the specific modules that are causing the issue. From the error message, it seems that cslam/global_descriptor_loop_closure_detection.py is one of the affected modules. You may need to look into the source code of these modules and find out how they are built and compiled.
If the modules are part of a larger project with a build system (e.g., CMake, Makefile, etc.), you'll need to follow the appropriate build procedures to recompile them with the updated dependencies (including the correct version of NumPy and pybind11). This could involve modifying build configuration files, running build commands, etc. The exact steps will depend on the specific build system used by the project.
Option 3: Check and Update cv2 Installation
Explanation: The issues with importing cv2 might be independent of the NumPy problem or could be related in some way. It's worth checking the installation of cv2 to ensure it's correct and up to date.
Steps:
First, uninstall the current version of cv2 if it's already installed. You can use pip to do this:
bash
复制
pip uninstall cv2
Then, reinstall cv2. The installation process might vary depending on your system and the specific requirements of your project. For example, if you're using OpenCV, you might need to install it from source or use a precompiled package. One common way to install OpenCV for Python is through the opencv-python package. You can install it using pip:
bash
复制
pip install opencv-python
After reinstalling cv2, try running your project again to see if the import issues are resolved.
In summary, the main issues seem to be related to NumPy version compatibility and potentially an incorrect installation or configuration of cv2. By trying one or more of the above solutions, you should hopefully be able to get your project running smoothly.

**Comment by lajoiepy** [comment]
Thank you @coisi-s for the detailed debug instructions. 
Compatibility of the python versions really depends on your system installation. The most common offenders are `cv2`, `numpy` and `networkx`. 
Swarm-SLAM is not really tied to a specific version as long as all your python packages are compatible with each other.
